# Preparing input data for DRMN

The regulatory features could be context-specific or context independent. For context-independent features, we use motif networks, by finding instances of motifs in promoter regions of genes.

![alt text](example_input/motif_small.png "Motif instances in gene promoter.")

For context-specific features, we can aggregate regulatory signals (different histone modification, or chromatin accessibility).

![alt text](example_input/signal_small.png "Aggregated signals in gene promoter.")

Furthermore, we can aggregate chromatin accessibility signals (e.g. DNase/ATAC-seq) signal in motif instances.

![alt text](example_input/qmotif_small.png "Q-Motif, aggregated signal in motif instances in gene promoter.")

We used our in-house softwares to aggregate signal in promoter regions or motif instances. 

### aggregateSignal

We used (aggregateSignal)[https://github.com/Roy-lab/aggregateSignalRegion_nonLog] to aggreage signal (ATAC-seq, DNase-seq, or histone modification) in promoter region of genes.

Briefly, we use bedtools to convert bam files to count files:
```
bedtools genomecov -ibam input.bam -bg -pc > output.counts
```
and aggregateSignal programs calculate coverage in input regions. 

The program can be used as:
```
./aggregateSignal promoters.txt mm10.fa.fai input.counts output.txt
```
where promoters.txt is a gff like file:
```
chr1	CONVERT	TSS_Capn8_ENSMUST00000192671	182562537	182567537	.	+	.	Capn8
chr2	CONVERT	TSS_Casc4_ENSMUST00000110586	121864469	121869469	.	+	.	Casc4
```
that contains the coordinates of each promoter (defined as +/- Xkb of TSS) and can be generated from ensembl gtf files.

The output.txt reports the signal per promoter of each gene:
```
Capn8	13.9519
Casc4	4.59291
```

### aggregateSignalMotifNet

We used (aggregateSignalMotifNet)[https://github.com/Roy-lab/aggregateSignalMotifNet] to aggreage signal (ATAC-seq, ...) in motif instances mapped to promoter region of genes.

The program can be used as:
```
./aggregateSignal motif_promoters.txt mm10.fa.fai input.counts output.txt
```
where motif_promoters.txt is a gff like file:
```
chr21	CONVERT	TSS_Gene2_TSS1	35198809	35198821	0	+	0	Motif2;Gene1
chr21	CONVERT	TSS_Gene2_TSS1	35144166	35144177	0	+	0	Motif2;Gene2
```
that contains the coordinates of motif instances in promoters. You can make this file by intersecting motif instances (as generated by motif finding program) and the promoter file (mentioned above) using bedtools, or you can use (PeakMapping)[https://github.com/Roy-lab/PeakMapping].

*Note:* The last column has both motif name and target gene name, separated by ";".

### Normalization

Furthermore, for each feature, we log transform and quantile normalize the values across cell lines/time points. 

### Formatting

Additionally, we add cell line/time point specific suffixes to gene names in order (so gene names will be specific to a cell line while regulators will be the same for all cell lines). 

In the feature file, the first line is number of regulators and number of genes in the file (tab-delim).
The rest of the file will be in 3 columns format (tab-delim):
```
feature g_cell  value
```
